presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java

```diff
@@ -0,0 +1,211 @@
+    {
+        return ImmutableSet.of("default");
+    }
+
+    private URI getPrometheusMetricsURI(PrometheusConnectorConfig config)
+            throws URISyntaxException
+    {
+        // endpoint to retrieve metric names from Prometheus
+        URI uri = config.getPrometheusURI();
+        return new URIBuilder()
+                .setScheme(uri.getScheme())
+                .setHost(uri.getAuthority())
+                .setPath(uri.getPath().concat(METRICS_ENDPOINT))
+                .build();
+    }
+
+    public Set<String> getTableNames(String schema)
+    {
+        requireNonNull(schema, "schema is null");
+        if (schema.equals("default")) {
+            String status = (String) tableSupplier.get().get("status");
+            if (status.equals("success")) {
+                List<String> tableNames = (List<String>) tableSupplier.get().get("data");
+                if (tableNames == null) {
+                    return ImmutableSet.of();
+                }
+                return tableNames.stream().collect(Collectors.toSet());
+            }
+            else {
+                //TODO this deals with success|error, there may be warnings to handle
+                return ImmutableSet.of();
+            }
+        }
+        return ImmutableSet.of();
+    }
+
+    public PrometheusTable getTable(String schema, String tableName)
+    {
+        requireNonNull(schema, "schema is null");
+        requireNonNull(tableName, "tableName is null");
+        if (schema.equals("default")) {
+            List<String> tableNames = (List<String>) tableSupplier.get().get("data");
+            if (tableNames == null) {
+                return null;
+            }
+            if (!tableNames.contains(tableName)) {
+                return null;
+            }
+            PrometheusTable table = new PrometheusTable(
+                    tableName,
+                    prometheusColumns);
+            pushedDown.ifPresent(pushedDownPredicate -> table.setPredicate(Optional.of(pushedDownPredicate)));
+            return table;
+        }
+        else {
+            return null;
+        }
+    }
+
+    private static Supplier<Map<String, Object>> metricsSupplier(final JsonCodec<Map<String, Object>> metricsCodec, final URI metadataUri)
+    {
+        return () -> {
+            try {
+                return lookupMetrics(metadataUri, metricsCodec);
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        };
+    }
+
+    private static Map<String, Object> lookupMetrics(URI metadataUri, JsonCodec<Map<String, Object>> metricsCodec)
+            throws IOException
+    {
+        String json = getHttpResponse(metadataUri).string();
+        Map<String, Object> metrics = metricsCodec.fromJson(json);
+
+        return metrics;
+    }
+
+    protected void setPredicate(Optional<TupleDomain<ColumnHandle>> pushedDown)
+    {
+        this.pushedDown = pushedDown;
```

Comment:
I guess this makes this class to be non thread safe. Is object of this class shared between queries?

Reply:
Eliminated as predicate is now stored in returned table handle in work done in b4e218c5